File: apngdrawable/src/main/java/com/dragon/drawable/apng/APngDrawable.java
package com.dragon.drawable.apng;

import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Rect;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.concurrent.ExecutorService;

import static android.graphics.PixelFormat.TRANSPARENT;

/**
 * @author chenjiulong
 */
public class APngDrawable extends Drawable implements Animatable, AnimationCallback {
    private APngHolder aPngHolder;
    private AnimationCallback animationCallback;

    private Runnable invalidateRunnable = () -> invalidateSelf();
    private Handler handler = new Handler();
    private boolean mStarting = false;
    private boolean alreadyEnded = false;

    public APngDrawable(APngHolder.IStream iStream, ExecutorService executorService) {
        aPngHolder = new APngHolder(iStream, executorService);
    }

    public APngDrawable(APngHolder aPngHolder) {
        this.aPngHolder = aPngHolder;
    }

    public void setAnimationCallback(AnimationCallback animationCallback) {
        this.animationCallback = animationCallback;
    }

    @Override
    public int getIntrinsicWidth() {
        return aPngHolder.getImageW();
    }

    @Override
    public int getIntrinsicHeight() {
        return aPngHolder.getImageH();
    }

    @Override
    public void setBounds(@NonNull Rect bounds) {
        super.setBounds(bounds);
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        if (mStarting) {
            mStarting = false;
            animationStart();
        }
        final long time = aPngHolder.draw(canvas);
        if (time >= 0) {
            scheduleSelf(invalidateRunnable, time);
        } else if ((time == -1 || time == -2) && !alreadyEnded) {
            alreadyEnded = true;
            animationEnd();
        }
    }

    @Override
    public void setAlpha(int alpha) {
        //do nothing
    }

    @Override
    public void setColorFilter(@Nullable ColorFilter colorFilter) {
        //do nothing
    }

    @Override
    public int getOpacity() {
        return TRANSPARENT;
    }


    @Override
    public boolean setVisible(boolean visible, boolean restart) {
        boolean ret = super.setVisible(visible, restart);
        return ret;
    }

    @Override
    public void start() {
        mStarting = true;
        alreadyEnded = false;
        aPngHolder.start();
        invalidateSelf();
    }

    @Override
    public void stop() {
        aPngHolder.stop();
    }

    @Override
    public boolean isRunning() {
        return aPngHolder.isRunning();
    }

    @Override
    public void animationStart() {
        if (animationCallback == null) {
            return;
        }
        handler.post(() -> {
            if (animationCallback != null) {
                animationCallback.animationStart();
            }
        });
    }

    @Override
    public void animationEnd() {
        if (animationCallback == null) {
            return;
        }
        handler.post(() -> {
            if (animationCallback != null) {
                animationCallback.animationEnd();
            }
        });
    }
}

File: APngLib/src/main/java/com/dragon/apnglib/APngDrawable.kt
package com.dragon.apnglib

import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.PixelFormat
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.os.Handler
import android.os.Looper
import android.os.Message
import kotlinx.coroutines.CoroutineScope
import java.io.InputStream
import java.util.*

/**
 * @author dragon
 */
class APngDrawable internal constructor(
        val file: String,
        sharedHolders: APngHolderPool?,
        scope: CoroutineScope,
        streamCreator: () -> InputStream
) : Drawable(), Animatable, Observer {

    private val holder: APngHolder = sharedHolders?.require(scope, file, streamCreator) ?: APngHolder(file, false, scope, streamCreator)

    var callback: AnimationCallback? = null

    private val handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            when (msg.obj) {
                APngHolder.PlayEvent.START -> {
                    callback?.onAnimationStart(this@APngDrawable)
                }
                APngHolder.PlayEvent.FRAME -> {
                    invalidateSelf()
                }
                APngHolder.PlayEvent.REPEAT -> {
                    callback?.onAnimationRepeat(this@APngDrawable)
                }
                APngHolder.PlayEvent.END -> {
                    callback?.onAnimationEnd(this@APngDrawable)
                }
                APngHolder.PlayEvent.CANCELED -> {
                    callback?.onAnimationCanceled(this@APngDrawable)
                }
            }
        }
    }

    init {
        holder.addObserver(this)
    }

    override fun draw(canvas: Canvas) {
        holder.draw(canvas)
    }

    override fun getIntrinsicWidth(): Int {
        return holder.columns
    }

    override fun getIntrinsicHeight(): Int {
        return holder.rows
    }

    override fun setVisible(visible: Boolean, restart: Boolean): Boolean {
        if (visible) holder.resume() else holder.pause()
        return super.setVisible(visible, restart)
    }

    override fun setAlpha(alpha: Int) {
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
    }

    override fun getOpacity(): Int {
        return PixelFormat.TRANSPARENT
    }

    override fun start() {
        holder.start()
    }

    override fun stop() {
        holder.stop()
    }

    override fun isRunning() = holder.isRunning

    override fun update(holder: Observable?, event: Any?) {
        handler.sendMessage(handler.obtainMessage().apply { obj = event })
    }

    fun release() {
        holder.deleteObserver(this)
    }

    fun finalize() {
        release()
    }
}